### ISSUE TITLE
Respect TreatTinyAsBoolean connection string option

### ISSUE THREAD

**Issue by AccessViolator** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#656]
### Background
Based mostly on [this article](https://www.xaprb.com/blog/2006/04/11/bit-values-in-mysql/) and the fact that bit(1) in MySql doesn't provide any space saving over tinyint(1) and how weirdly it's represented in various MySql tools I made a decision to use tinyint(1) for boolean values in my project. The article is from 2006, so some things mentioned there may have improved since, but still I feel that tinyint is a safer option - wouldn't want to run into some obscure issues in prodution.

### The issue
I use db first and scaffold models when I add tables, but I also generate test dbs off the models during some tests. With 2.0.1 I had to change sbyte to bool in scaffolded models. 2.1.1 [broke this at runtime](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/654). Current workaround is to use TreatTinyAsBoolean=false. I believe this is incorrect and having that option as true (default) should properly map tinyint(1) to bool when both scaffolding and generating tables (and work at runtime of course).

Yes I know that technically tinyint(1) can hold the same range of values as tinyint(4), but in this case I think it's irrelevant, as TreatTinyAsBoolean, as the [implementation in the underlying provider intends](https://github.com/mysql-net/MySqlConnector/issues/141), is specifically about treating (only) tinyint(1) (and should not affect other tinyint display sizes). And please don't bring ZeroFill into this :)

### Further technical details

MySQL version: 8.0.11
Operating system: Win 7 x64
Pomelo.EntityFrameworkCore.MySql version: 2.1.1

**Comment by caleblloyd** [comment]
Why change this now?  We've been operating fine with `bit(1)` from the start and I think it is working fine.

If we change it to `tinyint(1)` then we have to force people to use `TreatTinyAsBoolean=true` in their connection string.

**Comment by AccessViolator** [comment]
> Why change this now? We've been operating fine with bit(1) from the start and I think it is working fine.

It's only working fine for those who use bit(1) for bools. Yes, such users must be the majority by now. But do you want to limit users of this project to having to design their dbs in a particular way just because there's an old bug?
> If we change it to tinyint(1) then we have to force people to use TreatTinyAsBoolean=true in their connection string.

TreatTinyAsBoolean=true is the default setting, so it's already there.

As suggested in the opening post this should not affect how bit columns are mapped (and therefore majority of users): regardless of the TreatTinyAsBoolean setting bool with HasColumnType("bit(1)") should generate bit(1) column, bit(1) column is scaffolded to bool with HasColumnType("bit(1)"). Ditto for tinyint(3), tinyint(4).

Only tinyint(1) columns will be affected:

- With default setting (true) bool without HasColumnType or with HasColumnType("tinyint(1)") will generate tinyint(1). tinyint(1) column is scaffolded to bool with HasColumnType("tinyint(1)").
- With TreatTinyAsBoolean=false bool without HasColumnType generates bit(1) column, bool with HasColumnType("tinyint(1)") generates tinyint(1) column, tinyint(1) column is scaffolded to sbyte with HasColumnType("tinyint(1)").

So ideally we should be able to use bool with both HasColumnType("tinyint(1)") and HasColumnType("bit(1)") side by side without any runtime problems regardless of the setting. Setting would dictate what column is generated for bool without HasColumnType during scaffolding and whether tinyint(1) is scaffolded to bool or sbyte.

**Comment by Giantaxe** [comment]
How about adding a new option for it (like what got added for the characterset/collation aswell.)

A legacy switch basically,
```csharp
enum BooleanType
{
	Bit = 0,
	TinyInt = 1
}

/// Modified typemapping source
    // boolean
	private readonly MySqlBoolTypeMapping _bit              = new MySqlBoolTypeMapping()
	private readonly MySqlBoolTypeMapping _tinybool     = new MySqlBoolTypeMapping("tinyint")
		
	// constructor	
	_storeTypeMappings
		= new Dictionary<string, RelationalTypeMapping>(StringComparer.OrdinalIgnoreCase)
		{
			// boolean
			// { "bit", _bit },
			{ "bit", _options.BooleanType == BooleanType.Bit ? _bit : _tinybool },
```

ofcourse the MySQL Options class will have to be extended with the new property but wouldnt something like this work?

I guess it will only be an issue if you use both bit(1) and tinyint(1) in 1 and the same database, not sure what to do about that.

**Comment by caleblloyd** [comment]
@Giantaxe that is a good proposal. We will accept that as a PR.

Do you want to work on the PR @AccessViolator?

I can post some basic instructions here for which files to work in if you need me to.

**Comment by AccessViolator** [comment]
Not too keen to be honest :) And proposed solution feels a bit like a "crunch" around existing bug (no offense intended). With this approach to addressing problems behavior will continue to deviate from intended by MySqlConnector and this doesn't feel right to me. Current weird behavior of TreatTinyAsBoolean is nearly the opposite of what it should be, and it's not documented and very frustrating.

It must be possible with current architecture to support bools with both HasColumnType("bit(1)") and HasColumnType("tinyint(1)") in the same db as bool with HasColumnType("tinyint(1)") worked fine for me in 2.0.1, though I had to manually change sbyte to bool after scaffolding.

**Comment by caleblloyd** [comment]
We're not going to implement a breaking change. Also `bit(1)` is not a bug, it is the design.

If `tinyint(1)` is to be supported as a default behavior for bools, it must be done so behind a feature flag to maintain BC.

Otherwise you should be able to annotate every bool column with `HasColumnType("tinyint(1)")` today. Is that working properly?

**Comment by AccessViolator** [comment]
Well [#654](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/654) is a breaking change from 2.0.1. It's only working now with TreatTinyAsBoolean=fasle and only one way as tinyint(1) with this setting still scaffolds to sbyte.

**Comment by caleblloyd** [comment]
Ok, I reopened #654 to address the `HasColumnType("tinyint(1)")` annotation.

Scaffolding would probably require the feature flag discussed here, unless there is a way to define an alternate way for bool scaffolding.

**Comment by franva** [comment]
hi there, I am struggling to find out where to set TreatTinyAsBoolean. 
I am using Entity Framework Core database first approach.

Thanks

**Comment by AccessViolator** [comment]
connection string

**Comment by franva** [comment]
hi @AccessViolator 
I tried, still not working:
https://stackoverflow.com/questions/55652477/mysql-how-to-generate-property-type-bool-from-database-with-entity-framework-cor?noredirect=1#comment97994479_55652477

**Comment by a-patel** [comment]
@AccessViolator 
Any update, I am facing a similar issue using ADO.NET.

**Comment by lauxjpn** [comment]
While #875 should fix this issue for good, #866 contains workaround options (especially for anybody stuck on EF Core 2.2.x).

**Comment by lauxjpn** [comment]
Thanks @AccessViolator and everybody else for taking the time to report and discuss this issue.
This has been fixed in #875.
